File: sdk/typescript/src/namespaces/replays.ts
/**
 * Replays Namespace API
 *
 * Provides access to debate replays, event timelines, and replay visualization.
 */

/**
 * Replay event types
 */
export type ReplayEventType =
  | 'debate_start'
  | 'round_start'
  | 'round_end'
  | 'proposal'
  | 'critique'
  | 'revision'
  | 'vote'
  | 'consensus_check'
  | 'consensus_reached'
  | 'debate_end'
  | 'user_action'
  | 'system_event';

/**
 * Replay summary for listing
 */
export interface ReplaySummary {
  id: string;
  debate_id: string;
  task: string;
  agents: string[];
  total_rounds: number;
  total_events: number;
  duration_ms: number;
  result: 'consensus' | 'no_consensus' | 'timeout' | 'error';
  consensus_type?: string;
  created_at: string;
  size_bytes: number;
  has_video?: boolean;
}

/**
 * Replay event
 */
export interface ReplayEvent {
  id: string;
  type: ReplayEventType;
  timestamp: number;
  round?: number;
  agent?: string;
  content?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Full replay with events
 */
export interface Replay {
  id: string;
  debate_id: string;
  task: string;
  environment: Record<string, unknown>;
  protocol: Record<string, unknown>;
  agents: string[];
  events: ReplayEvent[];
  result: 'consensus' | 'no_consensus' | 'timeout' | 'error';
  consensus_proposal?: string;
  consensus_type?: string;
  total_tokens: number;
  total_cost?: number;
  created_at: string;
  metadata?: Record<string, unknown>;
}

/**
 * Evolution data showing how the debate progressed
 */
export interface EvolutionEntry {
  round: number;
  timestamp: number;
  proposals: Array<{
    agent: string;
    content: string;
    critique_count: number;
    revision_count: number;
    votes: number;
  }>;
  convergence_score: number;
  active_agents: string[];
}

/**
 * Replay fork for branching debates
 */
export interface ReplayFork {
  id: string;
  parent_replay_id: string;
  fork_point: number;
  fork_round: number;
  task_modification?: string;
  new_debate_id?: string;
  created_at: string;
}

/**
 * Interface for the internal client used by ReplaysAPI.
 */
interface ReplaysClientInterface {
  request<T>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Replays API namespace.
 *
 * Provides methods for accessing and analyzing debate replays:
 * - List and retrieve debate replays
 * - View event timelines
 * - Analyze debate evolution
 * - Fork debates from specific points
 * - Export replay visualizations
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // List recent replays
 * const { replays } = await client.replays.list({ limit: 10 });
 *
 * // Get a specific replay with all events
 * const replay = await client.replays.get('replay-id');
 *
 * // Get evolution data
 * const evolution = await client.replays.getEvolution('replay-id');
 * ```
 */
export class ReplaysAPI {
  constructor(private client: ReplaysClientInterface) {}

  /**
   * List available debate replays.
   */
  async list(options?: {
    agent?: string;
    result?: 'consensus' | 'no_consensus' | 'timeout' | 'error';
    since?: string;
    until?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ replays: ReplaySummary[]; total: number }> {
    return this.client.request('GET', '/api/replays', { params: options });
  }

  /**
   * Get a specific replay with all events.
   */
  async get(replayId: string): Promise<Replay> {
    return this.client.request('GET', `/api/replays/${replayId}`);
  }

  /**
   * Get just the events for a replay (lighter than full replay).
   */
  async getEvents(
    replayId: string,
    options?: {
      type?: ReplayEventType;
      agent?: string;
      round?: number;
      limit?: number;
      offset?: number;
    }
  ): Promise<{ events: ReplayEvent[]; total: number }> {
    return this.client.request('GET', `/api/replays/${replayId}/events`, { params: options });
  }

  /**
   * Get evolution data showing how the debate progressed.
   */
  async getEvolution(replayId: string): Promise<{ evolution: EvolutionEntry[] }> {
    return this.client.request('GET', `/api/replays/${replayId}/evolution`);
  }

  /**
   * Get HTML visualization of the replay.
   */
  async getHtml(replayId: string): Promise<string> {
    return this.client.request('GET', `/api/replays/${replayId}/html`);
  }

  /**
   * Fork a debate from a specific point to explore alternatives.
   */
  async fork(
    replayId: string,
    options: {
      fork_round?: number;
      fork_event_id?: string;
      task_modification?: string;
      new_agents?: string[];
    }
  ): Promise<ReplayFork> {
    return this.client.request('POST', `/api/replays/${replayId}/fork`, { json: options });
  }

  /**
   * List forks created from a replay.
   */
  async listForks(replayId: string): Promise<{ forks: ReplayFork[] }> {
    return this.client.request('GET', `/api/replays/${replayId}/forks`);
  }

  /**
   * Delete a replay (admin only).
   */
  async delete(replayId: string): Promise<{ success: boolean }> {
    return this.client.request('DELETE', `/api/replays/${replayId}`);
  }

  /**
   * Compare two replays to analyze differences.
   */
  async compare(
    replayId1: string,
    replayId2: string
  ): Promise<{
    replay_1: { id: string; task: string; result: string };
    replay_2: { id: string; task: string; result: string };
    similarities: string[];
    differences: string[];
    agent_overlap: string[];
    convergence_comparison: { replay_1: number; replay_2: number };
  }> {
    return this.client.request('GET', '/api/replays/compare', {
      params: { replay_id_1: replayId1, replay_id_2: replayId2 },
    });
  }

  /**
   * Get replay statistics.
   */
  async getStats(options?: {
    period?: '7d' | '30d' | '90d' | 'all';
  }): Promise<{
    total_replays: number;
    total_events: number;
    average_duration_ms: number;
    consensus_rate: number;
    average_rounds: number;
    by_result: Record<string, number>;
  }> {
    return this.client.request('GET', '/api/replays/stats', { params: options });
  }

  /**
   * Search replays by content.
   */
  async search(
    query: string,
    options?: {
      in_proposals?: boolean;
      in_critiques?: boolean;
      agent?: string;
      limit?: number;
    }
  ): Promise<{
    results: Array<{
      replay_id: string;
      task: string;
      matches: Array<{ event_id: string; content: string; highlight: string }>;
    }>;
  }> {
    return this.client.request('GET', '/api/replays/search', {
      params: { q: query, ...options },
    });
  }
}

File: sdk/python/aragora_sdk/namespaces/replays.py
"""
Replays namespace for debate replay management.

Provides API access to replay debates, analyze historical runs,
and review agent interactions over time.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class ReplaysAPI:
    """Synchronous replays API."""

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def search(
        self,
        query: str,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Search replays.

        Args:
            query: Search query.
            limit: Maximum results.
            offset: Pagination offset.

        Returns:
            Search results with replay matches.
        """
        params: dict[str, Any] = {"q": query, "limit": limit, "offset": offset}
        return self._client.request("GET", "/api/v1/replays/search", params=params)

    def get_stats(self) -> dict[str, Any]:
        """
        Get replay statistics.

        Returns:
            Replay statistics including counts and durations.
        """
        return self._client.request("GET", "/api/v1/replays/stats")

    def list(
        self,
        workspace_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available replays.

        Args:
            workspace_id: Filter by workspace
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of replays with pagination
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id:
            params["workspace_id"] = workspace_id

        return self._client.request("GET", "/api/v1/replays", params=params)

    def get(self, replay_id: str) -> dict[str, Any]:
        """
        Get a replay by ID.

        Args:
            replay_id: Replay identifier

        Returns:
            Replay details
        """
        return self._client.request("GET", f"/api/v1/replays/{replay_id}")

    def get_events(
        self,
        replay_id: str,
        event_type: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Get replay events.

        Args:
            replay_id: Replay identifier
            event_type: Filter by event type
            limit: Maximum events
            offset: Pagination offset

        Returns:
            Replay events
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if event_type:
            params["event_type"] = event_type

        return self._client.request("GET", f"/api/v1/replays/{replay_id}/events", params=params)

    def get_evolution(self, replay_id: str) -> dict[str, Any]:
        """
        Get the evolution of positions throughout the debate.

        Args:
            replay_id: Replay identifier

        Returns:
            Position evolution data
        """
        return self._client.request("GET", f"/api/v1/replays/{replay_id}/evolution")

    def fork(
        self,
        replay_id: str,
        from_event_index: int,
        new_input: str | None = None,
    ) -> dict[str, Any]:
        """
        Fork a replay from a specific point.

        Creates a new debate starting from a specific point in the replay,
        allowing exploration of alternative paths.

        Args:
            replay_id: Replay identifier
            from_event_index: Event index to fork from
            new_input: Optional new input for the fork

        Returns:
            Fork result with new debate_id
        """
        data: dict[str, Any] = {"from_event_index": from_event_index}
        if new_input:
            data["new_input"] = new_input

        return self._client.request("POST", f"/api/v1/replays/{replay_id}/fork", json=data)

    def compare(
        self,
        replay_id_1: str,
        replay_id_2: str,
    ) -> dict[str, Any]:
        """
        Compare two replays.

        Args:
            replay_id_1: First replay ID
            replay_id_2: Second replay ID

        Returns:
            Comparison analysis
        """
        return self._client.request(
            "GET",
            "/api/v1/replays/compare",
            params={"replay_id_1": replay_id_1, "replay_id_2": replay_id_2},
        )

    def delete(self, replay_id: str) -> dict[str, Any]:
        """
        Delete a replay.

        Args:
            replay_id: Replay identifier

        Returns:
            Deletion confirmation
        """
        return self._client.request("DELETE", f"/api/v1/replays/{replay_id}")


class AsyncReplaysAPI:
    """Asynchronous replays API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def search(
        self,
        query: str,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Search replays."""
        params: dict[str, Any] = {"q": query, "limit": limit, "offset": offset}
        return await self._client.request("GET", "/api/v1/replays/search", params=params)

    async def get_stats(self) -> dict[str, Any]:
        """Get replay statistics."""
        return await self._client.request("GET", "/api/v1/replays/stats")

    async def list(
        self,
        workspace_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List available replays."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id:
            params["workspace_id"] = workspace_id

        return await self._client.request("GET", "/api/v1/replays", params=params)

    async def get(self, replay_id: str) -> dict[str, Any]:
        """Get a replay by ID."""
        return await self._client.request("GET", f"/api/v1/replays/{replay_id}")

    async def get_events(
        self,
        replay_id: str,
        event_type: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Get replay events."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if event_type:
            params["event_type"] = event_type

        return await self._client.request(
            "GET", f"/api/v1/replays/{replay_id}/events", params=params
        )

    async def get_evolution(self, replay_id: str) -> dict[str, Any]:
        """Get the evolution of positions throughout the debate."""
        return await self._client.request("GET", f"/api/v1/replays/{replay_id}/evolution")

    async def fork(
        self,
        replay_id: str,
        from_event_index: int,
        new_input: str | None = None,
    ) -> dict[str, Any]:
        """Fork a replay from a specific point."""
        data: dict[str, Any] = {"from_event_index": from_event_index}
        if new_input:
            data["new_input"] = new_input

        return await self._client.request("POST", f"/api/v1/replays/{replay_id}/fork", json=data)

    async def compare(
        self,
        replay_id_1: str,
        replay_id_2: str,
    ) -> dict[str, Any]:
        """Compare two replays."""
        return await self._client.request(
            "GET",
            "/api/v1/replays/compare",
            params={"replay_id_1": replay_id_1, "replay_id_2": replay_id_2},
        )

    async def delete(self, replay_id: str) -> dict[str, Any]:
        """Delete a replay."""
        return await self._client.request("DELETE", f"/api/v1/replays/{replay_id}")
